### PR TITLE
Day 3

### PR DESCRIPTION
**Pull request by EchoAlice**

### CODE REVIEW

File: day3/part1.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        # Access each individual value in list.
+        for j in range(col):
+            if int(string[j]) == 0:
+                count_0[i] = count_0[i] + 1 
+            if int(string[j]) == 1:
+                count_1[i] = count_1[i] + 1
+        print("Count_0 ++: " + str(count_0[i]))
+        print("Count_1 ++: " + str(count_1[i]))
+
+        # How do I add 1 or 0 to a string to create a binary number?
```
Comment (EchoAlice) [inline day3/part1.py]: Delete this comment

File: day3/part1.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+if __name__ == "__main__":
```
Comment (EchoAlice) [inline day3/part1.py:1]: Holy shit.  You can start discussions on this thing

Comment (LeoTheMighty) [reply day3/part1.py:1]: Yeah it's really good for having debates or at least a back and forth on some decisions we had to make as programmers

File: day3/part1.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+if __name__ == "__main__":
+    
+    # Read text file into numbers within an array
+    with open("day3/input.txt") as values:
+         binary_input = values.readlines()
+
+    # Why is len(binary_input[0]) = 6?
+    col = len(binary_input[0]) - 1
```
Comment (LeoTheMighty) [inline day3/part1.py]: You should check and see what the last character is in this string, I'm sure you can run this in the debugger and find out why that's going on

Comment (LeoTheMighty) [inline day3/part1.py]: Also this could fail if `binary_input` is an empty list so I would usually just calculate the "width" in each for loop iteration based on the current array element you've accessed.

For instance:

```python
for i in range(len(array_2d)):
    row = array_2d[i]
    for j in range(len(row)):
        element = row[j]
```

File: day3/part1.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+if __name__ == "__main__":
+    
+    # Read text file into numbers within an array
+    with open("day3/input.txt") as values:
+         binary_input = values.readlines()
+
+    # Why is len(binary_input[0]) = 6?
+    col = len(binary_input[0]) - 1   
+    row = len(binary_input)
```
Comment (LeoTheMighty) [inline day3/part1.py]: Let's get some better names for these variables too: row and col sound like "items" to me and not like "lengths"/"measurements" even. This "row" variable specifically I think is how you're representing the height of the 2d array so you could call it `height`

File: day3/part1.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        if count_0 > count_1:
+            binary_input_gamma += str(0)
+            binary_input_epsilon += str(1)
+        # When ith digit in binary_input is a 1
+        if count_0 < count_1:
+            binary_input_gamma += str(1)
+            binary_input_epsilon += str(0)
+    #Convert
+    gamma_rate = int(binary_input_gamma, 2)
+    epsilon_rate =  int(binary_input_epsilon, 2)
```
Comment (LeoTheMighty) [inline day3/part1.py:27]: Haha python even has a binary number converter built in, isn't that amazing

File: day3/part1.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            if int(binary_input[j][i]) == 1:                
+                count_1 = count_1 + 1
+        # When ith digit in binary_input is a 0
+        if count_0 > count_1:
+            binary_input_gamma += str(0)
+            binary_input_epsilon += str(1)
+        # When ith digit in binary_input is a 1
+        if count_0 < count_1:
+            binary_input_gamma += str(1)
+            binary_input_epsilon += str(0)
```
Comment (LeoTheMighty) [inline day3/part1.py]: I like your logic, but it seems like too wordy to me and almost seems like it's repeating itself.

Let's try splitting these 7 lines into just 2. 

Look into python "ternary expressions" for a hint.

Comment (LeoTheMighty) [inline day3/part1.py]:
```suggestion
   what could,
   this be?
```

**Review by LeoTheMighty — CHANGES_REQUESTED** [review]
:)

File: day3/part1.py (hunk 3 above)
Comment (EchoAlice) [reply day3/part1.py]: I don't think this will work... I need to access the length of elements in each row in the outer loop, not the inner one

Comment (EchoAlice) [reply day3/part1.py]: It says the binary number is 13 digits long. When I print the 13th digit, nothing shows up as being printed

Comment (EchoAlice) [reply day3/part1.py]: The problem is that I need to access the width of the array first, so I can't have a variable as the element to access to find the length of each row.

for i in range(len(array_2d[0])):
    for j in range(len(array_2d));
        element = row[i][j]

**Comment by EchoAlice** [comment]
https://blog.ethereum.org/2015/11/15/merkling-in-ethereum/
Relatively easy to understand read.  I think you'll like it.
Talks about the power and efficiency of Merkle Trees
